docs/architecture/adr-005-relayer-v0-implementation.md

```diff
@@ -0,0 +1,179 @@
+structured.
+Here the focus is on the interfaces within the relayer, as well as the 
+interface between the relayer and a single chain.
+
+<img src="assets/relayer-v0-arch.jpg" width="500" alt="relayer v0 architecture"/>
+
+##### Levels of abstraction
+
+At the top of this diagram, there is a chain consisting of multiple full nodes.
+The deeper we go in this sketch, the closer we get to the user, or the 
+relayer CLI.
+To understand the relayer architecture intuitively, we can break down the 
+levels of abstraction as follows:
+
+###### 1. The actual chain, comprising a number of full nodes
+- This is the lowest level of abstraction, the farthest away from relayer 
+     users
+- The relayer communicates with a chain via three interfaces: 
+    - (i) the `LightClient` trait (handled via the supervisor for the 
+      production 
+     chain),
+    - (ii) the `Chain` trait (where the communication happens over the 
+  ABCI/gRPC interface primarily), and 
+    - (iii) an `EventMonitor` which subscribes to a full node and sends batches 
+  of events from that node to the chain runtime in the relayer. Currently, 
+  the relayer registers for `Tx` and `Block` notifications. It then extracts 
+  the IBC events from the Tx and generates a "NewBlock" event also for the
+  block. Note that the notification may include multiple IBC Events.
+     
+###### 2. The chain runtime
+- This is an intermediary layer, sitting between the relayer application and 
+  any chain(s)
+- It is universal for all possible chains, i.e., does _not_ contain any 
+     chain-specific code
+- Accepts as input requests from the application, in the form of 
+     [`ChainRequest`][3] via a crossbeam channel
+- Responds to the application via a crossbeam channel
+- Has objects which implement the three interfaces named above 
+  (`LightClient`, `Chain`, and `EventMonitor`) and orchestrates access to 
+  these objects as required by application requests
+
+###### 3. The relayer application
+- Communicates with the runtime via a `ChainHandle`, which contains the 
+      appropriate crossbeam sender and receiver channels to/from the runtime
+- Upon start-up, instantiates relayer-level objects in the following order: 
+  two `ForeignClient`s (one per chain), a `Connection` (which contains the 
+  two clients), a `Channel` (containing the connection), and on top of that 
+  a `Link`.
+
+##### Threads
+
+Each thread in this diagram is a separate box shaded in gray.
+There are four threads running: the `EventMonitor`, the `Supervisor`, the 
+`Runtime`, and the main application thread, called `V0Cmd`.
+
+[1]: https://github.com/cosmos/cosmos-sdk/pull/7993
+[2]: https://github.com/informalsystems/ibc-rs/blob/master/docs/architecture/adr-004-relayer-domain-decomposition.md#link
+[3]: https://github.com/informalsystems/ibc-rs/blob/379dd9812f6e7a42b9428f64eb52fe292d417476/relayer/src/chain/handle.rs#L51
+
+> TODO: Summary of pros and cons
```

Comment:
Note to self: might be useful to add here two stacktraces, as examples of what is the calling stack upon successfully submitting a simple transaction (create client) to either cosmos or mock chain (or both).